File: Maximal_Square_SpaceOptimized.cpp
// Time Complexity : O(m*n) traversing through the  dp matrix for all the values
// Space Complexity : O(n) for the dp vectors as we are using only 2 here. 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no issues as of now.. Learning


class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        // base condition
     if( matrix.size()== 0 || matrix[0].size() == 0) return 0;
        int m = matrix.size();
        int n = matrix[0].size();
        
      
      // instead of dp matrix using just 2 arrays   
        vector<int> dp( n+1,0); 
        vector<int> dp2( n+1,0); 
        

        // result initialization
        int result = 0;
        
        // traversing on the whole dp matrix
        for (int i = 1; i <=m; i++){
            
            for( int j =1; j <=n; j++){
                
                if(matrix[i-1][j-1] == '1'){
                       
                // dp condition or calculation/.. this will be done only if the matrix value is 1.
                    dp2[j] = min(dp2[j-1], min(dp[j], dp[j-1])) +1;
                    
                    // tracking the maximum value in martix
                    result = max(result,dp2[j]);
                }else dp2[j] = 0;
            }
            dp = dp2;
        }
        return result*result;
        
        
        
    }
};

File: MaximalSquareSpaceOptimized.java
/*

* Did this code successfully run on Leetcode : YES
* 
* Any problem you faced while coding this : NO
* 
* Time Complexity: O(m*n)
* 
* Space Complexity: O(n)
* 
*/

public class MaximalSquareSpaceOptimized {
    // int maxSquareSize = 0;

    // public int maximalSquare(char[][] matrix) {
    // int m = matrix.length;

    // int n = matrix[0].length;

    // int[] prev = new int[n];
    // int[] curr = new int[n];

    // for (int row = m - 1; row >= 0; row--) {
    // for (int col = n - 1; col >= 0; col--) {
    // if (matrix[row][col] == '1') {
    // int diag = 0, right = 0, down = 0;

    // if (col + 1 < n) {
    // diag = prev[col + 1];
    // }

    // down = prev[col];

    // if (col + 1 < n) {
    // right = curr[col + 1];
    // }

    // curr[col] = Math.min(Math.min(down, right), diag) + 1;

    // maxSquareSize = Math.max(maxSquareSize, curr[col]);

    // } else {
    // curr[col] = 0;
    // }
    // }
    // prev = curr.clone();
    // }

    // return maxSquareSize * maxSquareSize;
    // }

    int maxSquareSize = 0;

    public int maximalSquare(char[][] matrix) {
        int m = matrix.length;

        int n = matrix[0].length;

        int[] curr = new int[n];

        for (int row = m - 1; row >= 0; row--) {
            int diagDown = 0;

            for (int col = n - 1; col >= 0; col--) {
                int temp = curr[col];

                if (matrix[row][col] == '1') {
                    int right = 0, down = 0;

                    down = curr[col];

                    if (col + 1 < n) {
                        right = curr[col + 1];
                    }

                    curr[col] = Math.min(Math.min(down, right), diagDown) + 1;

                    maxSquareSize = Math.max(maxSquareSize, curr[col]);

                } else {
                    curr[col] = 0;
                }

                diagDown = temp;
            }
        }

        return maxSquareSize * maxSquareSize;
    }
}
